ucc: CUDA fixups

ConnorBaker opened the pull request:
Fixes https://github.com/NixOS/nixpkgs/issues/369226

- fix CUDA build (use cuda-compatible stdenv via `effectiveStdenv` "pattern")
- use `__structuredAttrs`
- use `strictDeps`
- switch to use of `fetchFromGitHub`'s `tag` argument
- build with `nccl`
- build with `nvml`
- remove deprecated use of `substituteInPlace`'s `--replace` flag
- remove `rec` keyword in favor of `finalAttrs` pattern

As future work, a number of these changes could be applied to UCX.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

SomeoneSerge commented on pkgs/by-name/uc/ucc/package.nix:
```diff
@@ -33,43 +56,55 @@ stdenv.mkDerivation rec {
     for comp in $(find src/components -name Makefile.am); do
-      substituteInPlace $comp \
-        --replace "/bin/bash" "${stdenv.shell}"
+      if ! grep -q "/bin/bash" "$comp"; then
+        continue
+      fi
```
Isn't this always false? I don't think /bin/bash ever exists in the sandbox (except when there's no sandbox)

ConnorBaker replied:
Some of the files don’t have `/bin/bash` in them so replacement would fail, so we skip them if they don’t match.

SomeoneSerge replied:
Ahh, I just forgot grep's order of arguments

SomeoneSerge commented on pkgs/by-name/uc/ucc/package.nix:
```diff
@@ -33,43 +56,55 @@ stdenv.mkDerivation rec {
   preConfigure =
+    # Fake libnvidia-ml.so (the real one is deployed impurely)
+    optionalString enableCuda ''
+      addToSearchPath LDFLAGS "-L${getLib cuda_nvml_dev}/lib/stubs"
```
H'm, I wonder who's responsible for picking up the real `LDFLAGS`. It's clear that `NIX_LDFLAGS_COMPILE` is picked up by the wrapper, but also we seem to prefer configuring higher-level tools (e.g. cmake) instead of the wrapper when possible

ConnorBaker replied:
I also tried exporting it through `LD_LIBRARY_PATH` but was unsuccessful.
I didn’t try modifying Nix’s equivalent `LDFLAGS` variables, would that be worth a try?
Stubs currently aren’t picked up by most hooks because they’re nested inside the `stubs` subdirectory of `lib`. In the work I’ve done out of tree I add symlinks in the stub output’s `lib` directory to the `stubs` subdirectory so they’re properly detected.

SomeoneSerge replied:
> I also tried exporting it through LD_LIBRARY_PATH but was unsuccessful.

That's different, that would need to be set at runtime. Like what we sometimes do with `makeWrapper`. `NIX_LDFLAGS_COMPILE` (Hmm I also see mentions of just `NIX_LDFLAGS` in the repo, which I didn't know exists) is essentially equivalent to `LDFLAGS` but I think it's translated into cli flags by the wrapper. So the 2 questions I have here are:

1. How do we choose between `{,NIX_}LDFLAGS{,_COMPILE}`
2. Why `addToSearchPath` instead of the initial value in `env` attributes

ConnorBaker replied:
I don't have an answer for 1) :(

For 2), I did it because I was unsure of whether setting `env.LDFLAGS` would clobber environment variables modified by setup hooks. Hopefully if there are any that touch it or use it, they're smart enough to not destroy original values.

SomeoneSerge replied:
I think not; `env` is like initial values, I think the "builder" program starts in an environment where these are already set: https://github.com/NixOS/nix/blob/442a2623e48357ff72c77bb11cf2cf06d94d2f90/src/libstore/unix/build/local-derivation-goal.cc#L2275

The setup-hooks are run by the "builder", inside the "build", and they normally use `addXXXX` in the expectation that there will be more hooks modifying the same variables before or after?

ConnorBaker replied:
Ah, I think I understand now!

As a secondary concern, are there setup hooks which don't use the `add*` family of bash functions? They could do bad things to environment variables which already have values.

SomeoneSerge commented on pkgs/by-name/uc/ucc/package.nix:
```diff
@@ -1,28 +1,51 @@
+in
+effectiveStdenv.mkDerivation (finalAttrs: {
+  __structuredAttrs = true;
+  # TODO: The M4 file requires a unified CUDA installation and doesn't support splayed outputs:
+  # https://github.com/openucx/ucc/blob/0c0fc21559835044ab107199e334f7157d6a0d3d/config/m4/cuda.m4
+  # Enabling strictDeps will break the build when CUDA is enabled.
+  strictDeps = false;
```
I don't quite understand how the former implies the latter?

ConnorBaker replied:
Good point! I need to reword it. 
Something is going on I don’t fully understand with `strictDeps`. When it is disabled, `configurePhase` succeeds and when it is enabled, `configurePhase` fails.
I believe it succeeds due to the non-separation of build/host dependencies, but need to find out whether that’s substantiated.

SomeoneSerge replied:
So either some of the buildInputs or some of the nativeBuildInputs are missing. I guess we don't have to debug and fix this right away, but maybe make the phrasing more explicit

markuskowa reviewed (commented):
Thanks! It builds again.
Two comments:
1. Using the `effectiveStdenv` solution needs to be re-implemented for every package that has CUDA support. I.e. ucc is only the first package in line to be fixed (?)
3. This PR contains a lot of different changes, all in one commit. It would be easier to read if formatting and pure cleanups would be in separate commits.

ConnorBaker commented on pkgs/by-name/uc/ucc/package.nix:
```diff
@@ -33,43 +56,55 @@ stdenv.mkDerivation rec {
     for comp in $(find src/components -name Makefile.am); do
-      substituteInPlace $comp \
-        --replace "/bin/bash" "${stdenv.shell}"
+      if ! grep -q "/bin/bash" "$comp"; then
+        continue
+      fi
+      substituteInPlace "$comp" \
+        --replace-fail \
+          "/bin/bash" \
+          "${effectiveStdenv.shell}"
```
The intent of the code is to replace occurrences of `/bin/bash`; issuing a warning when the substitution fails (because some of the files don’t contain `/bin/bash`) doesn’t make sense (there’s nothing to warn about because there’s nothing to fix). Instead, we should only patch files which do contain `/bin/bash` and fail when we are unable to.

I can rewrite the expression with `grep` instead of `find` to make that more clear if that would help!

SomeoneSerge replied:
I think the current impl is fairly clear? A shorter but definitely obscure option might be to `--replace-warn ... 2>/dev/null` xD

ConnorBaker replied:
The current implementation uses the deprecated `--replace` argument :(

markuskowa replied:
There is also `--replace-quiet` which should do the job.

ConnorBaker commented:
> Thanks! It builds again. Two comments:
> 
> 1. Using the `effectiveStdenv` solution needs to be re-implemented for every package that has CUDA support. I.e. ucc is only the first package in line to be fixed (?)

As of right now that’s correct — packages which compile CUDA code during their build should use the `effectiveStdenv` pattern.

> 2. This PR contains a lot of different changes, all in one commit. It would be easier to read if formatting and pure cleanups would be in separate commits.

Will do!

ConnorBaker committed:
ucc: general fixups

ConnorBaker committed:
ucc: CUDA fixups

NovaViper commented:
Gave the PR a test (copied the package.nix file into my flake and overlayed the package) and it successfully builds!

ConnorBaker commented:
Addressed most comments and force-pushed.

@SomeoneSerge your question

> How do we choose between {,NIX_}LDFLAGS{,_COMPILE}

is an excellent one and I think it deserves its own issue. Would you mind creating one for it?

SomeoneSerge commented:
> ...deserves its own issue. Would you mind creating one for it?

I can. OTOH we've been creating backlog issues one after another, faster than we close issues... I'm not saying "not every thought is worth writing down" but at least in the format of github issues, feels a bit pointless. It would make sense in Logseq or a wiki though, where you have backlinks

As for the question itself, maybe instead of an issue it needs a discourse thread with at least some analysis of the existing practices, i.e. it's just more work that needs getting done

SomeoneSerge commented on pkgs/by-name/uc/ucc/package.nix:
```diff
@@ -1,28 +1,52 @@
+effectiveStdenv.mkDerivation (finalAttrs: {
+  __structuredAttrs = true;
+  # TODO(@connorbaker):
+  # When strictDeps is enabled, `cuda_nvcc` is required as the argument to `--with-cuda` in `configureFlags` or else
+  # configurePhase fails with `checking for cuda_runtime.h... no`.
+  # This is odd, especially given `cuda_runtime.h` is provided by `cuda_cudart.dev`, which is already in `buildInputs`.
```
Maybe they compile a test program to "detect" cuda_runtime.h? 🙃

SomeoneSerge reviewed (approved):
didn't test runtime/buildtime but I read the diff, lgtm

osbm commented:
Thanks for the awesome work @ConnorBaker . You are a life saver.

NovaViper commented:
Hey is there anything else that needs to be done in order for the PR to be merged? I've been using it for a bit now and the changes made has fixed all of the issues related with cuda failing with trying to build ucc

osbm commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 369956`

---
### `x86_64-linux`
<details>
  <summary>:fast_forward: 28 packages marked as broken and skipped:</summary>
  <ul>
    <li>froide</li>
    <li>froide.dist</li>
    <li>octavePackages.sparsersb</li>
    <li>python312Packages.bambi</li>
    <li>python312Packages.bambi.dist</li>
    <li>python312Packages.froide</li>
    <li>python312Packages.froide.dist</li>
    <li>python312Packages.hoomd-blue</li>
    <li>python312Packages.pymc</li>
    <li>python312Packages.pymc.dist</li>
    <li>python312Packages.python-mapnik</li>
    <li>python312Packages.python-mapnik.dist</li>
    <li>python312Packages.rio-tiler</li>
    <li>python312Packages.rio-tiler.dist</li>
    <li>python312Packages.shimmy</li>
    <li>python312Packages.shimmy.dist</li>
    <li>python312Packages.worldengine</li>
    <li>python312Packages.worldengine.dist</li>
    <li>python312Packages.wrf-python</li>
    <li>python312Packages.wrf-python.dist</li>
    <li>python313Packages.hoomd-blue</li>
    <li>python313Packages.python-mapnik</li>
    <li>python313Packages.python-mapnik.dist</li>
    <li>python313Packages.worldengine</li>
    <li>python313Packages.worldengine.dist</li>
    <li>t-rex</li>
    <li>worldengine-cli</li>
    <li>worldengine-cli.dist</li>
  </ul>
</details>
<details>
  <summary>:x: 44 packages failed to build:</summary>
  <ul>
    <li>cloudcompare</li>
    <li>elmerfem</li>
    <li>globalarrays</li>
    <li>gnudatalanguage</li>
    <li>librsb</li>
    <li>libtensorflow (libtensorflow.python)</li>
    <li>mni_autoreg</li>
    <li>netcdf-mpi</li>
    <li>neuron-full (python312Packages.neuronpy)</li>
    <li>octopus</li>
    <li>otb</li>
    <li>pnfft</li>
    <li>pnfft.dev</li>
    <li>pymol</li>
    <li>pymol.dist</li>
    <li>python312Packages.datashader</li>
    <li>python312Packages.datashader.dist</li>
    <li>python312Packages.fenics</li>
    <li>python312Packages.fenics.dist</li>
    <li>python312Packages.gstools</li>
    <li>python312Packages.gstools.dist</li>
    <li>python312Packages.itk</li>
    <li>python312Packages.meshio</li>
    <li>python312Packages.meshio.dist</li>
    <li>python312Packages.pydicom-seg</li>
    <li>python312Packages.pydicom-seg.dist</li>
    <li>python312Packages.pyprecice</li>
    <li>python312Packages.pyprecice.dist</li>
    <li>python312Packages.qcodes-contrib-drivers</li>
    <li>python312Packages.qcodes-contrib-drivers.dist</li>
    <li>python312Packages.sfepy</li>
    <li>python312Packages.sfepy.dist</li>
    <li>python313Packages.fenics</li>
    <li>python313Packages.fenics.dist</li>
    <li>python313Packages.gstools</li>
    <li>python313Packages.gstools.dist</li>
    <li>python313Packages.itk</li>
    <li>python313Packages.meshio</li>
    <li>python313Packages.meshio.dist</li>
    <li>python313Packages.pydicom-seg</li>
    <li>python313Packages.pydicom-seg.dist</li>
    <li>python313Packages.pyprecice</li>
    <li>python313Packages.pyprecice.dist</li>
    <li>rocmPackages_5.rocprofiler</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 373 packages built:</summary>
  <ul>
    <li>aligator</li>
    <li>aligator.doc</li>
    <li>ants</li>
    <li>apacheHttpdPackages.mod_tile (apacheHttpdPackages_2_4.mod_tile)</li>
    <li>arpack-mpi</li>
    <li>bicgl</li>
    <li>bicpl</li>
    <li>bonmin</li>
    <li>c3d</li>
    <li>casadi</li>
    <li>cdo</li>
    <li>conglomerate</li>
    <li>costa</li>
    <li>cp2k</li>
    <li>crocoddyl</li>
    <li>crocoddyl.doc</li>
    <li>dbcsr</li>
    <li>dl-poly-classic-mpi</li>
    <li>dtcmp</li>
    <li>dtcmp.dev</li>
    <li>ebtks</li>
    <li>eccodes</li>
    <li>elastix</li>
    <li>elpa</li>
    <li>elpa.dev</li>
    <li>elpa.doc</li>
    <li>elpa.man</li>
    <li>entwine</li>
    <li>ezminc</li>
    <li>fftwMpi</li>
    <li>fftwMpi.dev</li>
    <li>fftwMpi.info</li>
    <li>fftwMpi.man</li>
    <li>freecad</li>
    <li>freecad-wayland</li>
    <li>gdal (python312Packages.gdal)</li>
    <li>getdp</li>
    <li>gmt</li>
    <li>gplates</li>
    <li>grass</li>
    <li>gromacsCudaMpi</li>
    <li>gromacsCudaMpi.dev</li>
    <li>gromacsCudaMpi.man</li>
    <li>gromacsDoubleMpi</li>
    <li>gromacsDoubleMpi.dev</li>
    <li>gromacsDoubleMpi.man</li>
    <li>gromacsMpi</li>
    <li>gromacsMpi.dev</li>
    <li>gromacsMpi.man</li>
    <li>haskellPackages.hgdal</li>
    <li>haskellPackages.hgdal.doc</li>
    <li>haskellPackages.mpi-hs-binary</li>
    <li>haskellPackages.mpi-hs-binary.doc</li>
    <li>haskellPackages.mpi-hs-cereal</li>
    <li>haskellPackages.mpi-hs-cereal.doc</li>
    <li>haskellPackages.mpi-hs-store</li>
    <li>haskellPackages.mpi-hs-store.doc</li>
    <li>hdf5-mpi</li>
    <li>hdf5-mpi.bin</li>
    <li>hdf5-mpi.dev</li>
    <li>highfive-mpi</li>
    <li>hp2p</li>
    <li>hpcg</li>
    <li>hpl</li>
    <li>hypre</li>
    <li>inormalize</li>
    <li>intensity-normalization (python312Packages.intensity-normalization)</li>
    <li>intensity-normalization.dist (python312Packages.intensity-normalization.dist)</li>
    <li>ior</li>
    <li>ipopt</li>
    <li>itk (itk_5)</li>
    <li>itk_5_2</li>
    <li>labplot</li>
    <li>lammps-mpi</li>
    <li>libcircle</li>
    <li>libcircle.dev</li>
    <li>libmbd</li>
    <li>libminc</li>
    <li>libvdwxc</li>
    <li>lwgrp</li>
    <li>lwgrp.dev</li>
    <li>mapcache</li>
    <li>mapnik</li>
    <li>mapproxy</li>
    <li>mapproxy.dist</li>
    <li>mapserver</li>
    <li>merkaartor</li>
    <li>migrate</li>
    <li>mim-solvers</li>
    <li>minc_tools</li>
    <li>minc_widgets</li>
    <li>mirtk</li>
    <li>mpi (openmpi)</li>
    <li>mpi.dev (openmpi.dev)</li>
    <li>mpi.man (openmpi.man)</li>
    <li>mpifileutils</li>
    <li>mpifileutils.dev</li>
    <li>mpifileutils.man</li>
    <li>mrtrix</li>
    <li>mumps</li>
    <li>mumps_par</li>
    <li>mysql-workbench</li>
    <li>n3</li>
    <li>nco</li>
    <li>ncview</li>
    <li>ndcurves</li>
    <li>ndcurves.doc</li>
    <li>nest-mpi</li>
    <li>netcdf</li>
    <li>netcdfcxx4</li>
    <li>netcdffortran</li>
    <li>neuron-mpi</li>
    <li>nwchem</li>
    <li>octavePackages.mapping</li>
    <li>octavePackages.ncarray</li>
    <li>octavePackages.netcdf</li>
    <li>oobicpl</li>
    <li>openorienteering-mapper</li>
    <li>openturns</li>
    <li>ovito</li>
    <li>p4est</li>
    <li>p4est-dbg</li>
    <li>p4est-sc</li>
    <li>p4est-sc-dbg</li>
    <li>pagmo2</li>
    <li>paraview</li>
    <li>parmetis</li>
    <li>parmmg</li>
    <li>pdal</li>
    <li>pdi</li>
    <li>perl538Packages.Tirex</li>
    <li>perl538Packages.Tirex.devdoc</li>
    <li>perl540Packages.Tirex</li>
    <li>perl540Packages.Tirex.devdoc</li>
    <li>petsc</li>
    <li>pfft</li>
    <li>pfft.dev</li>
    <li>pinocchio</li>
    <li>pinocchio.doc</li>
    <li>precice</li>
    <li>proxsuite-nlp</li>
    <li>proxsuite-nlp.doc</li>
    <li>python312Packages.aerosandbox</li>
    <li>python312Packages.aerosandbox.dist</li>
    <li>python312Packages.aligator</li>
    <li>python312Packages.aligator.doc</li>
    <li>python312Packages.arviz</li>
    <li>python312Packages.arviz.dist</li>
    <li>python312Packages.baselines</li>
    <li>python312Packages.baselines.dist</li>
    <li>python312Packages.boltztrap2</li>
    <li>python312Packages.boltztrap2.dist</li>
    <li>python312Packages.bsuite</li>
    <li>python312Packages.bsuite.dist</li>
    <li>python312Packages.cartopy</li>
    <li>python312Packages.cartopy.dist</li>
    <li>python312Packages.casadi</li>
    <li>python312Packages.corner</li>
    <li>python312Packages.corner.dist</li>
    <li>python312Packages.crocoddyl</li>
    <li>python312Packages.crocoddyl.doc</li>
    <li>python312Packages.cyipopt</li>
    <li>python312Packages.cyipopt.dist</li>
    <li>python312Packages.dask-mpi</li>
    <li>python312Packages.dask-mpi.dist</li>
    <li>python312Packages.django-bootstrap4</li>
    <li>python312Packages.django-bootstrap4.dist</li>
    <li>python312Packages.django-bootstrap5</li>
    <li>python312Packages.django-bootstrap5.dist</li>
    <li>python312Packages.drf-extra-fields</li>
    <li>python312Packages.drf-extra-fields.dist</li>
    <li>python312Packages.eccodes</li>
    <li>python312Packages.example-robot-data</li>
    <li>python312Packages.example-robot-data.doc</li>
    <li>python312Packages.fiona</li>
    <li>python312Packages.fiona.dist</li>
    <li>python312Packages.flowmc</li>
    <li>python312Packages.flowmc.dist</li>
    <li>python312Packages.folium</li>
    <li>python312Packages.folium.dist</li>
    <li>python312Packages.geoarrow-pandas</li>
    <li>python312Packages.geoarrow-pandas.dist</li>
    <li>python312Packages.geoarrow-pyarrow</li>
    <li>python312Packages.geoarrow-pyarrow.dist</li>
    <li>python312Packages.geodatasets</li>
    <li>python312Packages.geodatasets.dist</li>
    <li>python312Packages.geopandas</li>
    <li>python312Packages.geopandas.dist</li>
    <li>python312Packages.geoparquet</li>
    <li>python312Packages.geoparquet.dist</li>
    <li>python312Packages.gpaw</li>
    <li>python312Packages.gpaw.dist</li>
    <li>python312Packages.h5netcdf</li>
    <li>python312Packages.h5netcdf.dist</li>
    <li>python312Packages.h5py-mpi</li>
    <li>python312Packages.h5py-mpi.dist</li>
    <li>python312Packages.inequality</li>
    <li>python312Packages.inequality.dist</li>
    <li>python312Packages.libpysal</li>
    <li>python312Packages.libpysal.dist</li>
    <li>python312Packages.mapclassify</li>
    <li>python312Packages.mapclassify.dist</li>
    <li>python312Packages.medpy</li>
    <li>python312Packages.medpy.dist</li>
    <li>python312Packages.medvol</li>
    <li>python312Packages.medvol.dist</li>
    <li>python312Packages.meep</li>
    <li>python312Packages.mim-solvers</li>
    <li>python312Packages.momepy</li>
    <li>python312Packages.momepy.dist</li>
    <li>python312Packages.morecantile</li>
    <li>python312Packages.morecantile.dist</li>
    <li>python312Packages.mpi4py</li>
    <li>python312Packages.mpi4py.dist</li>
    <li>python312Packages.msticpy</li>
    <li>python312Packages.msticpy.dist</li>
    <li>python312Packages.napari-nifti</li>
    <li>python312Packages.napari-nifti.dist</li>
    <li>python312Packages.ndcurves</li>
    <li>python312Packages.ndcurves.doc</li>
    <li>python312Packages.nest</li>
    <li>python312Packages.netbox-documents</li>
    <li>python312Packages.netbox-documents.dist</li>
    <li>python312Packages.netcdf4</li>
    <li>python312Packages.netcdf4.dist</li>
    <li>python312Packages.neuralfoil</li>
    <li>python312Packages.neuralfoil.dist</li>
    <li>python312Packages.nifty8</li>
    <li>python312Packages.nifty8.dist</li>
    <li>python312Packages.nutpie</li>
    <li>python312Packages.nutpie.dist</li>
    <li>python312Packages.openturns</li>
    <li>python312Packages.osmnx</li>
    <li>python312Packages.osmnx.dist</li>
    <li>python312Packages.pinocchio</li>
    <li>python312Packages.pinocchio.doc</li>
    <li>python312Packages.plotnine</li>
    <li>python312Packages.plotnine.dist</li>
    <li>python312Packages.proxsuite-nlp</li>
    <li>python312Packages.proxsuite-nlp.doc</li>
    <li>python312Packages.pygmo</li>
    <li>python312Packages.pygmt</li>
    <li>python312Packages.pygmt.dist</li>
    <li>python312Packages.pymedio</li>
    <li>python312Packages.pymedio.dist</li>
    <li>python312Packages.pyogrio</li>
    <li>python312Packages.pyogrio.dist</li>
    <li>python312Packages.pyradiomics</li>
    <li>python312Packages.pyradiomics.dist</li>
    <li>python312Packages.qcodes</li>
    <li>python312Packages.qcodes.dist</li>
    <li>python312Packages.rasterio</li>
    <li>python312Packages.rasterio.dist</li>
    <li>python312Packages.rioxarray</li>
    <li>python312Packages.rioxarray.dist</li>
    <li>python312Packages.simpleitk</li>
    <li>python312Packages.simpleitk.dist</li>
    <li>python312Packages.sirius</li>
    <li>python312Packages.sirius.dev</li>
    <li>python312Packages.torchio</li>
    <li>python312Packages.torchio.dist</li>
    <li>python312Packages.tsid</li>
    <li>python312Packages.tsid.doc</li>
    <li>python312Packages.wktutils</li>
    <li>python312Packages.wktutils.dist</li>
    <li>python313Packages.aerosandbox</li>
    <li>python313Packages.aerosandbox.dist</li>
    <li>python313Packages.aligator</li>
    <li>python313Packages.aligator.doc</li>
    <li>python313Packages.boltztrap2</li>
    <li>python313Packages.boltztrap2.dist</li>
    <li>python313Packages.casadi</li>
    <li>python313Packages.crocoddyl</li>
    <li>python313Packages.crocoddyl.doc</li>
    <li>python313Packages.cyipopt</li>
    <li>python313Packages.cyipopt.dist</li>
    <li>python313Packages.dask-mpi</li>
    <li>python313Packages.dask-mpi.dist</li>
    <li>python313Packages.django-bootstrap4</li>
    <li>python313Packages.django-bootstrap4.dist</li>
    <li>python313Packages.django-bootstrap5</li>
    <li>python313Packages.django-bootstrap5.dist</li>
    <li>python313Packages.drf-extra-fields</li>
    <li>python313Packages.drf-extra-fields.dist</li>
    <li>python313Packages.eccodes</li>
    <li>python313Packages.example-robot-data</li>
    <li>python313Packages.example-robot-data.doc</li>
    <li>python313Packages.gdal</li>
    <li>python313Packages.gpaw</li>
    <li>python313Packages.gpaw.dist</li>
    <li>python313Packages.h5netcdf</li>
    <li>python313Packages.h5netcdf.dist</li>
    <li>python313Packages.h5py-mpi</li>
    <li>python313Packages.h5py-mpi.dist</li>
    <li>python313Packages.intensity-normalization</li>
    <li>python313Packages.intensity-normalization.dist</li>
    <li>python313Packages.medpy</li>
    <li>python313Packages.medpy.dist</li>
    <li>python313Packages.medvol</li>
    <li>python313Packages.medvol.dist</li>
    <li>python313Packages.meep</li>
    <li>python313Packages.mim-solvers</li>
    <li>python313Packages.mpi4py</li>
    <li>python313Packages.mpi4py.dist</li>
    <li>python313Packages.napari-nifti</li>
    <li>python313Packages.napari-nifti.dist</li>
    <li>python313Packages.ndcurves</li>
    <li>python313Packages.ndcurves.doc</li>
    <li>python313Packages.nest</li>
    <li>python313Packages.netbox-documents</li>
    <li>python313Packages.netbox-documents.dist</li>
    <li>python313Packages.netcdf4</li>
    <li>python313Packages.netcdf4.dist</li>
    <li>python313Packages.neuralfoil</li>
    <li>python313Packages.neuralfoil.dist</li>
    <li>python313Packages.nifty8</li>
    <li>python313Packages.nifty8.dist</li>
    <li>python313Packages.openturns</li>
    <li>python313Packages.pinocchio</li>
    <li>python313Packages.pinocchio.doc</li>
    <li>python313Packages.proxsuite-nlp</li>
    <li>python313Packages.proxsuite-nlp.doc</li>
    <li>python313Packages.pygmt</li>
    <li>python313Packages.pygmt.dist</li>
    <li>python313Packages.pymedio</li>
    <li>python313Packages.pymedio.dist</li>
    <li>python313Packages.pyogrio</li>
    <li>python313Packages.pyogrio.dist</li>
    <li>python313Packages.pyradiomics</li>
    <li>python313Packages.pyradiomics.dist</li>
    <li>python313Packages.simpleitk</li>
    <li>python313Packages.simpleitk.dist</li>
    <li>python313Packages.sirius</li>
    <li>python313Packages.sirius.dev</li>
    <li>python313Packages.tsid</li>
    <li>python313Packages.tsid.doc</li>
    <li>pytrainer</li>
    <li>pytrainer.dist</li>
    <li>qgis</li>
    <li>qgis-ltr</li>
    <li>qmapshack</li>
    <li>quantum-espresso</li>
    <li>raxml-mpi</li>
    <li>rocmPackages.rocalution (rocmPackages_6.rocalution)</li>
    <li>rocmPackages.rocprofiler (rocmPackages_6.rocprofiler)</li>
    <li>rocmPackages_5.rocalution</li>
    <li>saga</li>
    <li>scalapack</li>
    <li>scalapack.dev</li>
    <li>scotch</li>
    <li>scotch.bin</li>
    <li>scotch.dev</li>
    <li>siesta-mpi</li>
    <li>simpleitk</li>
    <li>sirius</li>
    <li>sirius.dev</li>
    <li>spfft</li>
    <li>spla</li>
    <li>spla.dev</li>
    <li>sumo</li>
    <li>survex</li>
    <li>therion</li>
    <li>tmpi</li>
    <li>trilinos-mpi</li>
    <li>tsid</li>
    <li>tsid.doc</li>
    <li>tunnelx</li>
    <li>ucc</li>
    <li>ucc.dev</li>
    <li>udig</li>
    <li>vpv</li>
    <li>xyce-parallel</li>
    <li>xyce-parallel.doc</li>
  </ul>
</details>

osbm commented:
This review took 8 hours.

@NovaViper when i saw your comment the last hour i thought this PR would be merged right before anyone seeing the output 😂

I think some of these broken packages i failed to build doesnt have issues open, should i open them in order to notify the maintainers?

NovaViper commented:
👀 That looks like something with Python.. I hadn't updated in a week (the 9th to be exact), will give it a update later to see if I run into the same issues

But this is what my flake's built against currently
```
❯ nix-info -m
 - system: `"x86_64-linux"`
 - host os: `Linux 6.12.8, NixOS, 25.05 (Warbler), 25.05.20250108.bffc22e`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.25.3`
 - nixpkgs: `/nix/store/v0g0bxsd5gw6k0jz2855f8h7l1218925-source`

```

osbm commented:
I believe that all of these packages were not broken because of the ucc package. But i am not %100 sure.

The package that i desperately need the most is python3Packages.simpleitk which this PR actually fixes.

markuskowa reviewed (approved):
I am still not convinced that using the `effectiveStdenv` trick is a good solution. It is rather messy, since we need to put that in every package that has a CUDA option. I think this should be solved somewhere at the top level, so that not every package has to handle the issue.
However, for now let's fix the problem.

markuskowa commented:
@GrahamcOfBorg build ucc

SomeoneSerge commented:
> I am still not convinced that using the effectiveStdenv trick is a good solution

👍🏻 It most definitely isn't. A thought that occurred to me just now: what lead us to the `effectiveStdenv` (anti-)pattern is that we were allowing local overrides using `.override { enableCuda = true; }`. The function of this flag correlates with the function of the `stdenv` argument, and it's this entanglement that actually causes the mess.

Maybe `.override { cudaSupport = true; }` should have been `.override { stdenv = cudaStdenv; }` in the first place. All `lib.optionals cudaSupport` should have been `lib.optionals stdenv.isCuda` or `stdenv.isNVCC`. We could remove (move to `stdenv`) all cuda- or hip-related flags, but the easiest way to do that is to replace the actual default `stdenv` for all packages, triggering even more rebuilds than we do now.

Accidentally, this would also make it easier to move `config.cudaSupport` into `crossSystem` and introduce a `pkgsCross.cuda`.

That said, I'm uncomfortable with rebuilding the entire package set with the old compiler as opposed to using the old compiler selectively when and only when a derivation uses nvcc. Thoughts? @ConnorBaker @samuela @markuskowa?

ConnorBaker commented:
I don’t think there is a good solution — using a custom stdenv requires either universally changing the default stdenv and comes with the problems you mentioned or marking derivations in such a way that we could surgically replace the stdenv for those which use CUDA.

The alternative approach is to wrap NVCC and use setup hooks to ensure it has a compatible version of a compiler and that it doesn’t leak out. But then LTO won’t work because you’d be trying to optimize across binaries produced by different compiler versions.

For what it’s worth, I’ve taken the second approach of wrapping NVCC here (https://github.com/ConnorBaker/cuda-packages) and am trying to gauge the impact that causes (e.g., broken LTO).
